### ISSUE TITLE
How to use this library for sending metrics to AppInsights?

### ISSUE THREAD

**Issue by govenkat-ms** [frigus02/opentelemetry-application-insights#39]
Can this library be used for sending metrics to AppInsights?

Thanks

**Comment by frigus02** [comment]
Okay, I think the API for traces also supports metrics. It's just a different data model.

Sadly, it looks like neither the official Python ([source](https://github.com/Azure/azure-sdk-for-python/tree/main/sdk/monitor/azure-monitor-opentelemetry-exporter/azure/monitor/opentelemetry/exporter/export)) nor JavaScript ([source](https://github.com/Azure/azure-sdk-for-js/tree/main/sdk/monitor/monitor-opentelemetry-exporter/src/export)) exporters support metrics. This means we need to figure out the mapping from OpenTelemetry model to AppInsights model ourselves.

If I see this correctly, this is the AppInsights model ([swagger](https://github.com/Azure/azure-rest-api-specs/blob/e47988f3ccaa90681f11cce59c25c638ff305692/specification/applicationinsights/data-plane/Monitor.Exporters/preview/2020-09-15_Preview/swagger.json#L439)):

```json
{
  "name": "Microsoft.ApplicationInsights.Metrics",
  "time": "2009-06-15T13:45:30.0000000Z",
  "data": {
    "baseType": "MetricsData",
    "baseData": {
      "metrics": [
        {
          "ver": 2,
          "name": "banana",
          "value": 4.2,
          "kind": "Measurement OR Aggregation",
          "count": 2,
          "min": 1.0,
          "max": 10.0,
          "stdDev": 0.5
        }
      ]
    }
  }
}
```

**Comment by frigus02** [comment]
Looks like there is also another API for metrics called "Azure Metrics":

- [Docs](https://docs.microsoft.com/en-us/azure/azure-monitor/essentials/metrics-store-custom-rest-api)
- [Swagger schema](https://github.com/Azure/azure-rest-api-specs/blob/88e7838a09868a51de3894114355c75929847a46/specification/monitor/data-plane/preview/2018-09-01-preview/metricsCreate_API.json)

This is looks harder to use because of the OAuth authentication.

I don't have any experience using metrics in AppInsights. Do you know which one is preferred right now, @govenkat-ms?

**Comment by frigus02** [comment]
If we would implement this based on the AppInsights API, we would need to map whatever the [different aggregators](https://docs.rs/opentelemetry/0.16.0/opentelemetry/sdk/export/metrics/trait.Aggregator.html#implementors) give us to the MetricsData model.

- ArrayAggregator: probably a list of "Measurement" metrics each with just a value
- LastValueAggregator: probably just a single "Measurement" metric with just a value
- MinMaxSumCountAggregator: a single "Aggregation" metric with min, max, count and value=(sum/count)
- DdSketchAggregator: same as MinMaxSumCountAggregator
- SumAggregator: a single "Aggregation" metric with value=sum
- HistogramAggregator: could report a single "Aggregation" metric with count and value=(sum/count). Not sure how to encode the buckets, though. Maybe just ignore them?

I imagine this list is likely subject to change as the OpenTelemetry metrics API finalizes and the Rust implementation catches up. But kinds of aggregations should not change drastically.

**Comment by govenkat-ms** [comment]
I don't believe there's a preference to either the docs or the swagger.

**Comment by frigus02** [comment]
I have an intial implementation of metrics, which seems to be working. Still needs some more testing of the different aggregation types but this looks promising.

**Comment by frigus02** [comment]
Sorry this took so long. I couldn't find the time to finish this.

Anyway. Metrics support is now in and released in version 0.18.0. The API to setup the exporter is still a bit rough. But I think that can be improved in the future.
